quickwit-indexing/src/source/kafka_source.rs

```diff
@@ -296,86 +304,78 @@ impl Source for KafkaSource {
+                message_res = self.consumer.recv() => {
+                    let message = match message_res {
+                        Ok(message) => message,
+                        Err(KafkaError::PartitionEOF(partition_id)) => {
+                            self.state.num_active_partitions -= 1;
+                            info!(
+                                topic = %self.topic,
+                                partition_id = ?partition_id,
+                                num_active_partitions = ?self.state.num_active_partitions,
+                                "Reached end of partition."
+                            );
+                            continue;
                         }
-                        batch_num_bytes += message.payload_len() as u64;
-                        self.state.num_bytes_processed += message.payload_len() as u64;
-                        self.state.num_messages_processed += 1;
-
-                        let partition_id = self
-                            .state
-                            .assigned_partition_ids
-                            .get(&message.partition())
-                            .ok_or_else(|| {
-                                anyhow::anyhow!(
-                                    "Received message from unassigned partition `{}`. Assigned partitions: \
-                                     `{{{}}}`.",
-                                    message.partition(),
-                                    self.state.assigned_partition_ids.keys().join(", "),
-                                )
-                            })?
-                            .clone();
-                        let current_position = Position::from(message.offset());
-                        let previous_position = self
-                            .state
-                            .current_positions
-                            .insert(message.partition(), current_position.clone())
-                            .unwrap_or_else(|| previous_position_for_offset(message.offset()));
-                        checkpoint_delta
-                            .record_partition_delta(partition_id, previous_position, current_position)
-                            .context("Failed to record partition delta.")?;
-
-                        if batch_num_bytes >= TARGET_BATCH_NUM_BYTES {
-                            break;
-                        }
-                        ctx.record_progress();
+                        // FIXME: This is assuming that Kafka errors are not recoverable, it may not be the
+                        // case.
+                        Err(err) => return Err(ActorExitStatus::from(anyhow::anyhow!(err))),
+                    };
+                    if let Some(doc) = parse_message_payload(&message) {
+                        docs.push(doc);
                     } else {
-                        // TODO: this is probably not what should happen here.
```

Comment:
Removing the comment since a source actor behaves precisely as suggested. The source actor is a spinning actor: when we exit `emit_batches`, the actor performs internal bookkeeping and then calls the method again.

Reply:
Yeah, I agree. It was the partition.eof behavior that was making me question this.